### PR TITLE
Support generating metrics from spans

### LINKED ISSUE #357: Add support for metrics from traces
There is an incoming PR to the opentelemetry-collector-contrib repo that will allow the generation of metrics from traces:

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2012

Once this is merged let's vendor and expose this config to our users.

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

#### PR Description 

This pr adds the support for `spanmetricsprocessor` to generate metrics from spans.
It will create two metrics, one is a counter called `calls` and another one is a histogram called `latency`.
For its configuration and usage, please check https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.23.0/processor/spanmetricsprocessor/README.md.

`spanmetricsprocessor` requires a metrics exporter to expose the generated metrics. Here I think it might be better to allow users to configure the exporter. Now I only add prometheus exporter support.

I haven't updated the doc because I am not sure about my implementation and the configuration. If the impl looks good, I can add the document later.

#### Which issue(s) this PR fixes 

Fixes #357

#### PR Checklist

- [ ] CHANGELOG updated 
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by yeya24** [comment]
I already tested this manually with the config below. After traces are pushed, I can get the generated metrics from `localhost:8889/metrics`.

``` yaml
server:
  http_listen_port: 8081
tempo:
  configs:
  - name: tempo
    push_config:
      ...

    spanmetrics:
      latency_histogram_buckets: [ 2ms, 6ms, 10ms, 100ms, 250ms ]
      dimensions:
        - name: http.method
          default: GET
        - name: http.status_code
      metrics_exporter:
        name: prometheus
        config:
          endpoint: "0.0.0.0:8889"
          namespace: promexample

    receivers:
      jaeger:
        protocols:
          grpc:
          thrift_binary:
          thrift_compact:
          thrift_http:
      zipkin:
      otlp:
        protocols:
          http:
          grpc:
      opencensus:
```

File: pkg/tempo/config.go
Comment (mapno) [inline pkg/tempo/config.go:159]: Should we allow multiple exporters to be configured?

**Review by mapno — COMMENTED** [review]
Thank you for your contribution! Great work!
The changes are in line with what I had in mind.  I think we can move forward with the approach.  I just left some minor comments.

I also tried the example you provided while adding the prometheus exporter to `scrape_configs` and they get reported, so very cool!

File: pkg/tempo/dummyreceiver/factory.go
Comment (mapno) [inline pkg/tempo/dummyreceiver/factory.go:14]: What do you think of naming it `noop` instead?

**Review by mapno — COMMENTED** [review]
On second thought, would it make more sense to configure a prometheus exporter by default?  The main use case will be to have the agent self-scrape the exported prometheus metrics and have them sent to the configured `remote_write`.  I think it makes it a simpler feature for the users to enable (i.e. less configuration).

The future goal will be to hook this metrics directly to the agent's remote write, instead of having to scrape them locally.

File: pkg/tempo/dummyreceiver/factory.go
Comment (yeya24) [reply pkg/tempo/dummyreceiver/factory.go:14]: Yeah, I am good with this.

File: pkg/tempo/config.go
Comment (yeya24) [reply pkg/tempo/config.go:304]: > On second thought, would it make more sense to configure a prometheus exporter by default? The main use case will be to have the agent self-scrape the exported prometheus metrics and have them sent to the configured `remote_write`. I think it makes it a simpler feature for the users to enable (i.e. less configuration).
> 
> The future goal will be to hook this metrics directly to the agent's remote write, instead of having to scrape them locally.

File: pkg/tempo/config.go
Comment (yeya24) [reply pkg/tempo/config.go:167]: Good idea.

**Comment by yeya24** [comment]
> would it make more sense to configure a prometheus exporter by default
If we want to have a prometheus exporter by default and we still need to support the remote write, then what would the configuration file look like? Can you please give me an example?

> The future goal will be to hook this metrics directly to the agent's remote write, instead of having to scrape them locally.
Do you mean we want to implement a new exporter for the existing Prometheus remote write part? Without using the otlp prometehus remote write exporter?

File: pkg/tempo/config.go
Comment (yeya24) [reply pkg/tempo/config.go:304]: We need to define the exporter configuration first. If we just want to support Prometheus exporter then having a default namespace is fine.

**Comment by rfratto** [comment]
> The future goal will be to hook this metrics directly to the agent's remote write, instead of having to scrape them locally.

My $0.02: I understand this is the future goal, but I personally think we should *not* have this use the Prometheus Exporter, and would rather keep the current approach of having the Agent scrape itself. The Prometheus Exporter doesn't have a WAL, and I would feel weird having some of the Agent's metrics be more vulnerable to data loss. 

I think moving towards the future goal should be done in a follow up PR (especially because gut feeling tells me it's not going to be as simple as it sounds) and we should avoid temporarily using the Prometheus exporter. 

Just a suggestion though! Ultimately I want to leave the call up to you and/or @joe-elliott to decide what to do for this iteration of the feature.

File: docs/configuration-reference.md
Comment (mapno) [inline docs/configuration-reference.md:2070]: This option does not exist in the version of the OTel collector we're using

```suggestion
```

**Review by mapno — APPROVED** [review]
LGTM. Left just 3 minor comments. Thanks again for the contribution!

**Comment by yeya24** [comment]
@mapno Hello, I have updated my pr.

File: CHANGELOG.md
Comment (rfratto) [inline CHANGELOG.md:24]:
```suggestion
- [FEATURE] (beta) Support generating metrics and exposing them via a Prometheus exporter
```

**Review by rfratto — APPROVED** [review]
This LGTM, but I'm going to leave it up to @mapno to merge since I've mostly been disengaged.

cc @joe-elliott in case you want to give a final review

File: docs/configuration-reference.md
Comment (joe-elliott) [inline docs/configuration-reference.md:2066]: spelling: `prometehus` => `prometheus`

File: docs/configuration-reference.md
Comment (joe-elliott) [inline docs/configuration-reference.md:2052]: it's not clear to me if this is a remote write configuration or a scrape endpoint configuration. i'm guessing this just opens an http server on a port and exposes the metrics at `/metrics` ?

either way can we get a little more verbiage to clarify?

**Review by joe-elliott — COMMENTED** [review]
one nit and one request. fine with merging this otherwise.

thank you for the contribution!

File: docs/configuration-reference.md
Comment (yeya24) [reply docs/configuration-reference.md]: Exactly. Can you take a another look?

**Review by joe-elliott — APPROVED** [review]

**Pull request merged**
